File: 序号/322.零钱兑换.kt
class Solution {
    fun coinChange(coins: IntArray, amount: Int): Int {
        // 定义dp[i]为凑i钱需要最小硬笔数量
        val dp = IntArray(amount+1){-1}
        dp[0]=0
        for (i in 0..amount){
            for (coin in coins){
                if (i-coin>=0 && dp[i-coin]!=-1) {
                    if(dp[i]==-1) dp[i]=dp[i-coin]+1
                    else if(dp[i-coin]+1 < dp[i]){
                        dp[i]=dp[i-coin]+1
                    }
                }
            }
        }
        return dp[amount]
    }
}

File: 序号/322. 零钱兑换.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @cache
        def dfs(amount):
            if amount<0:return -1
            if amount==0:return 0
            # mn = min(dfs(amount-c) for c in coins)
            mn=-1
            for c in coins:
                p = dfs(amount-c)
                if p<0:continue
                if mn==-1:mn=p
                elif p<mn:mn=p
            # print(amount,mn)
            return -1 if mn==-1 else mn+1
        return dfs(amount)